boards/arm/mec172xevb_assy6906/pinmux.c

```diff
@@ -9,9 +9,105 @@
 #include <kernel.h>
 #include <drivers/pinmux.h>
 
-#include "soc.h"
+#include <soc.h>
 
-/*
- * If PINMUX will be used this is where board_pinmux_init should be implemented
- * and added to SYS_INIT.
- */
+#define XEC_UART_0_REGS							\
+	((struct uart_regs *)DT_REG_ADDR(DT_NODELABEL(uart0)))
+
+#define XEC_UART_1_REGS							\
+	((struct uart_regs *)DT_REG_ADDR(DT_NODELABEL(uart1)))
+
+struct pinmux_ports_t {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_000_036), okay)
+	const struct device *porta;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_040_076), okay)
+	const struct device *portb;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_100_136), okay)
+	const struct device *portc;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_140_176), okay)
+	const struct device *portd;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_200_236), okay)
+	const struct device *porte;
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pinmux_240_276), okay)
+	const struct device *portf;
+#endif
```

Comment:
So I do think you can use the foreach macros. Can you upload the CL where you tried it and we can help debug what you're doing wrong here?

Reply:
@jackrosenthal Latest push contains commented out DT_FOREACH experiments that generate errors.  When I look at other chips for example Atmel and RISC-V I see pinmux under soc with pinmux-someid: pinmux@some-address {};  For Atmel samd2x pinmux and gpio have the same addresses but the DT compiler does not complain.  If I try what Atmel did I get a DTC warning about overlapping addresses.  The HW GPIO pin multiplex control is in the same register as most of the other GPIO configuration.  This means two drivers touching the same registers.  This may explain why the original author of pinmux implemented the way it is.  Most other SoC's have separate HW registers for pin multiplexing and other pin features.